rfcs/20211207-softmax-v2/README.md

```diff
@@ -0,0 +1,192 @@
+doc) does not have it. None of other frameworks are supporting this parameter.
+Following documentation references were used:
+[Tensorflow](https://www.tensorflow.org/api_docs/python/tf/nn/softmax),
+[PyTorch](https://pytorch.org/docs/stable/generated/torch.nn.Softmax.html),
+[OpenVINO](https://docs.openvino.ai/latest/openvino_docs_ops_activation_SoftMax_1.html).
+In its core, it would divide the data by `temperature`, see
+[MXNet documentation](https://mxnet.apache.org/versions/master/api/python/docs/api/npx/generated/mxnet.npx.softmax.html)
+for details. Same can be achieved if the preceding layer would apply
+`binary_div` post-op with `mask` of `0`. Due to the lack of broader adoption and
+existence of simple work around, such feature will not be added to oneDNN
+softmax primitive.
+
+Another long-lasting request is so-called `masked` softmax. It comes from
+Transformer-based models when Decoder applies Encoder-Decoder Attention layer.
+At this point the machine is not allowed to see future tokens of the sentence,
+that's why this "mask" should hide the future starting from the processed token.
+Mask makes softmax input data shorter.
+
+Due to the nice property of softmax that data on the left end of range does not
+contribute much or at all (depending on the range) to final answer and likely
+vanishes, there is a trick to perform this optimization on the user side, which
+will be explained shortly. Thus, usually matmul precedes by softmax. The mask
+can be incorporated into bias which would fill specific spots with `-inf` or
+`INT32_MIN` values and fuse its addition into matmul. Then softmax will handle a
+full, non-masked, length but those points will not contribute to the final
+answer and will not spoil the accuracy of the model. Such trick allows to handle
+a desired behavior much faster than softmax could do it.
+
+Solution via softmax primitive has API, integration, and implementation
+complexity. That's why masked feature won't be added to softmax primitive as
+well.
+
+## Proposal
+
+### Option 1
+
+There are no options to cover different memory descriptors and algorithms other
+than creating a new operation descriptor, `dnnl_softmax_v2_desc_t`. It will be
+extended over previous one with `dnnl_alg_kind_t alg_kind`,
+`dnnl_memory_desc_t dst_desc` and `dnnl_memory_desc_t diff_dst_desc`.
+`data_desc` analogues will be replaced with `src_desc` correspondently.
+Implementation details will be similar to pooling_v2.
+
+#### C API
+
+```c
+/// dnnl_types.h
+
+/// Kinds of primitives. Used to implement a way to extend the library with new
+/// primitives without changing the ABI.
+typedef enum {
+    ...
+    /// A softmax version 2 primitive (softmax with destination memory
+    /// descriptor and algorithm kind).
+    dnnl_softmax_v2,
+    ...
+} dnnl_primitive_kind_t;
+
+/// Kinds of algorithms.
+typedef enum {
+    ...
+    /// Softmax
+    dnnl_softmax_alg,
```

Comment:
Speaking of the open questions, using `alg` suffix seems odd to me as we don't actually use it for any other algorithm kinds (except undef).

We can use algorithm names from [cuDNN](https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnSoftmaxAlgorithm_t): `dnnl_softamx_fast` and `dnnl_softmax_log`.
This will make things easier for those who want to migrate their cuDNN based code to oneDNN.

Reply:
Yeah, this is a good suggestion, I'll use it, thank you. Didn't come to me to check with cudnn. Though I think we can name it softmax_fast_and_accurate :D

Reply:
Fast and Furious :D

Reply:
So, what we actually have in oneDNN is an equivalent of `CUDNN_SOFTMAX_ACCURATE`? If I'm not mistaken we search for the maximum value and then we do scaling using that value.

Reply:
Yes, that was my intention to name it like this.

Reply:
Updated alg names, thank you for the tip.